src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleMetadata.cs

```diff
@@ -14,42 +14,32 @@ public static partial class JsonSerializer
         internal static readonly JsonEncodedText s_metadataRef = JsonEncodedText.Encode("$ref", encoder: null);
         internal static readonly JsonEncodedText s_metadataValues = JsonEncodedText.Encode("$values", encoder: null);
 
-        internal static MetadataPropertyName GetResolvedReferenceHandling(
-            JsonConverter converter,
-            object value,
-            ref WriteStack state,
-            out string? referenceId)
-        {
-            if (!converter.CanHaveIdMetadata || converter.TypeToConvert.IsValueType)
-            {
-                referenceId = default;
-                return MetadataPropertyName.NoMetadata;
-            }
-
-            if (state.ReferenceResolver.TryGetOrAddReferenceOnSerialize(value, out referenceId))
-            {
-                return MetadataPropertyName.Ref;
-            }
-
-            return MetadataPropertyName.Id;
-        }
-
         internal static MetadataPropertyName WriteReferenceForObject(
             JsonConverter jsonConverter,
             object currentValue,
             ref WriteStack state,
             Utf8JsonWriter writer)
         {
-            MetadataPropertyName metadataToWrite = GetResolvedReferenceHandling(jsonConverter, currentValue, ref state, out string? referenceId);
+            MetadataPropertyName metadataToWrite;
 
-            if (metadataToWrite == MetadataPropertyName.Ref)
+            // If the jsonConverter supports immutable dictionaries or value types, don't write any metadata
+            if (!jsonConverter.CanHaveIdMetadata || jsonConverter.TypeToConvert.IsValueType)
+            {
+                metadataToWrite = MetadataPropertyName.NoMetadata;
+            }
+            else if (state.ReferenceResolver.TryGetOrAddReferenceOnSerialize(currentValue, out string referenceId))
             {
-                writer.WriteString(s_metadataRef, referenceId!);
+                Debug.Assert(referenceId != null);
+                writer.WriteString(s_metadataRef, referenceId);
                 writer.WriteEndObject();
+                metadataToWrite = MetadataPropertyName.Ref;
             }
-            else if (metadataToWrite == MetadataPropertyName.Id)
+            else
             {
-                writer.WriteString(s_metadataId, referenceId!);
+                // TryGetOrAddReferenceOnSerialize is guaranteed to not return null.
```

Comment:
Is this comment necessary? The purpose of nullable reference types is you shouldn't need a comment like this, given the signature says "out string" rather than "out string?"

Reply:
I added it along with the `Debug.Assert` to make it explicit that `referenceId` shouldn't be null even if `TryGetOrAddReferenceOnSerialize` accidentally set `referenceId` to `null!`. Otherwise, it's a bug in the implementation detail.

I am fine removing the comment though. Should I remove the `Debug.Assert` as well?

Reply:
> Should I remove the Debug.Assert as well?

The comment seemed superfluous.  I'm fine with the Debug.Assert if you think it adds value.